<h3>Задача семинара</h3>

<p>Мы будем решать задачу, о которой говорили на лекции &mdash;&nbsp;реализуем&nbsp;механизм подсказок (<em>suggest</em>) категории по заголовку.</p>

<p>Представьте, что вы хотите разместить объявление о продаже шин. Чтобы положить своё объявление в нужную категорию, вам необходимо выбрать одну из 10 верхнеуровневых категорий, одну из 5 на следющем уровне, одну из 12 на следующем уровне, одну из 5 на последнем уровне.</p>

<p>Даже если варианты правильных&nbsp;категорий&nbsp;на каждом уровне очевидны, процесс подачи объявления лучше упростить.</p>

<div style="text-align:center">
<figure class="image" style="display:inline-block"><img alt="" height="259" src="http://labelimages.avito.ru/mlcourse_week3_without_suggest.png" width="563" />
<figcaption>Пример поиска нужной категории без подсказок</figcaption>
</figure>
</div>

<p>&nbsp;Упрощённый процесс подачи объявлений может выглядеть следующим образом. Пользователь заполняет заголовок объявления, затем выбирает один из предложенных вариантов категорий, который больше всего подходит его объявлению.</p>

<div style="text-align:center">
<figure class="image" style="display:inline-block"><img alt="" height="324" src="http://labelimages.avito.ru/mlcourse_week3_with_suggest.png" width="534" />
<figcaption>Работа suggest при публикации объявления</figcaption>
</figure>
</div>

<p>Задача &mdash; определить категорию объявления, используя заголовок. Всего в выборке будет 54 категории.</p>


<h3><strong>Многоклассовая классификация</strong></h3>

<p>Для решения задачи будем использовать модель логистической регрессии – линейная модель для бинарной классификации. Как её использовать для решения задачи классификации на 54 класса? Оказывается, любой бинарный классификатор, который способен оценивать меру принадлежности объектов к одному из 2 классов, можно использовать для многоклассовой классификации. </p>

<p>Мнгоклассовая классификация на основе бинарного классификатора может быть реализована с помощью одной из двух стратегий:</p>

<ul>
	<li>Один против всех (<em>One-vs-Rest</em> или <em>One-vs-All</em>);</li>
	<li>One-vs-One.</li>
</ul>

<p> </p>

<p><strong>One-vs-All </strong></p>

<p>В стратегии <em>One-vs-All</em> строятся <span class="math-tex">\(k\)</span> классификаторов, отделяющих каждый класс от остальных. Например, для 3 классов нужно обучить 3 классификатора:</p>

<div style="text-align: center;">
<figure class="image" style=""><img alt="" height="436" name="Снимок экрана 2021-08-03 в 19.26.56.png" src="https://ucarecdn.com/2abe5812-3f6e-42cf-97a4-a35a6de58c6d/" width="513">
<figcaption>Многоклассовая классфикация по стратегии OVR</figcaption>
</figure>
</div>

<p>Помимо эффективности построения (необходимы только <span class="math-tex">\(k\)</span> классификаторов), одним из преимуществ этого подхода является его интерпретируемость. Поскольку для каждого класса обучался свой классификатор, можно анализировать каждый классификатор в отдельности, чтобы понять, почему модель приняла такое решение.</p>

<p><em>Предсказания на тестовой выборке для OVR.</em></p>

<p>Для объектов из тестовой выборки предсказывается класс, получивший максимальную вероятность в результате предсказаний <span class="math-tex">\(k\)</span> классификаторов:</p>

<p style="text-align: center;"><span class="math-tex">\(\hat{y} = \underset{k}{argmax} \ a_{k}(x),\)</span></p>

<p>где <span class="math-tex">\(a_k(x) -\)</span> предсказание модели бинарной классификации для класса <span class="math-tex">\(k\)</span>.  Выражение<em> <em><em><em><em><em><em><em><span class="math-tex">\(\argmax\)</span></em></em></em></em></em></em></em></em> означает выбор <span class="math-tex">\(k\)</span>, при котором значение <span class="math-tex">\(a_{k}(x)\)</span> максимально.</p>

<p><em>На практике обычно используют OVR стратегию или алгоритмы с нативной поддержкой многоклассовой классификации – нейронные сети или алгоритмы, которые мы будем изучать дальше в этом курсе (RandomForest, KNN).</em></p>

<p> </p>

<p><strong>One-vs-One (опционально)</strong></p>

<p>В стратегии <em>One-vs-One </em>обучается один классификатор для каждой пары классов. При этом каждый классификатор обучается только на объектах, относящихся к этим классам: </p>

<div style="text-align: center;">
<figure class="image" style=""><img alt="" height="551" name="Снимок экрана 2021-08-01 в 23.43.15.png" src="https://ucarecdn.com/b6a950d0-a787-49de-b6cd-9f2faadb91c3/" width="637">
<figcaption>Многклассовая классификация стратегией OVO</figcaption>
</figure>
</div>

<p>Таким образом, требуется построить квадратичное число классификаторов <span class="math-tex">\(\frac{k(k-1)}{2}\)</span>.</p>

<p>При больших <span class="math-tex">\(k\)</span> может потребоваться много времени на обучение. Однако этот метод может быть полезен для алгоритмов, которые плохо масштабируются с увеличением размера выборки, так как каждая отдельная задача обучения классификатора включает в себя только объекты, принадлежащие двум выбранным классам. </p>

<p><em>Предсказания на тестовой выборке для OVO.</em></p>

<p>Для объектов из тестовой выборки все <span class="math-tex">\(\frac{k(k-1)}{2}\)</span> классификаторов делают предсказания, а финальный ответ получается в результате «жёсткого» голосования:</p>

<p style="text-align: center;"><span class="math-tex">\(\hat{y} = \underset{k}{argmax} \sum_{m \neq k} [a_{km}(x) = 1],\)</span></p>

<p>где  <span class="math-tex">\(a_{km}(x) -\)</span>классификатор с позитивным классом <span class="math-tex">\(k\)</span> и негативным классом <span class="math-tex">\(m\)</span>.  Жёстким оно называется потому, что никак не учитывается, насколько уверены были классификаторы в своих предсказаниях.</p>

<p>Если несколько классов для объекта получили одинаковое количество голосов или нужно определить не только класс, но и меру принадлежности, производится «мягкое» голосование: выигрывает тот класс в котором суммарно больше уверены классификаторы.</p>


## EDA

<p>Начнем работу с данными с импорта нужных библиотек:</p>

In [1]:
from google.colab import drive
drive.mount('/content/drive')
way = "/content/drive/MyDrive/Colab_Notebooks/aaa_ml/datasets/"

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import numpy as np
import pandas as pd

from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

<p>Прочитаем <a href="https://stepik.org/media/attachments/lesson/537384/suggest_train.csv" rel="noopener noreferrer nofollow">данные</a>:</p>


In [3]:
data = pd.read_csv(way+'suggest_train.csv', index_col=0)
data.head(3)

title  category_id  \
item_id                                                                   
335330            Раствор. Товарный бетон от производителя           15   
137480                                    Детская кроватка           16   
242690   Кровать -качалка с ящиком для вещей, два полож...           34   

                                                      name  
item_id                                                     
335330   Для дома и дачи|Ремонт и строительство|Стройма...  
137480   Для дома и дачи|Мебель и интерьер|Кровати, див...  
242690   Личные вещи|Товары для детей и игрушки|Детская...

Видим, что данные упорядочены по категориям. Это может привести к ошибке при разбиении данных. Перемешаем данные:

In [4]:
data = data.sample(frac=1).reset_index(drop=True)

Проверим, правильно ли закодированы категории:

In [5]:
np.sort(data.category_id.unique())

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53])

In [6]:
data.category_id.min(), data.category_id.max()

(0, 53)

Категории закодированы правильно.

**Баланс классов**

Посмотрим на распределение классов. Большой дисбаланс может вызвать проблемы модели. Например, в топ-3 по вероятности будут попадать нерелевантные категории просто потому, что в датасете их намного больше, чем остальных.

In [7]:
data.groupby('category_id').size().describe()

count     54.000000
mean     370.370370
std       33.178361
min      309.000000
25%      346.000000
50%      361.000000
75%      398.750000
max      447.000000
dtype: float64

Видим, что минимальное количество примеров на категорию 309, а максимальное 447. Скорее всего, проблемы не возникнет.

**Примеры заголовков**

Посмотрим на примеры топ-3 наиболее и наименее популярных категорий:

In [8]:
data[data.name.isin(data.name.value_counts()[:3].index)].drop_duplicates(subset='name')

title  category_id  \
5               Подушка "овечья шерсть"           27   
10                  Горка под тв Оливия           23   
81  Нуга Бест Турманиевый Жилет MHP-100           37   

                                                 name  
5   Для дома и дачи|Мебель и интерьер|Текстиль и к...  
10           Для дома и дачи|Мебель и интерьер|Другое  
81  Личные вещи|Красота и здоровье|Приборы и аксес...

Наименее популярные:

In [9]:
data[data.name.isin(data.name.value_counts()[-3:].index)].drop_duplicates(subset='name')

title  category_id  \
12                       Речной жемчуг           40   
32   HP 15-d050sr Intel DualCore N2810            1   
227           Продам телевизор Samsung            5   

                                                  name  
12      Личные вещи|Часы и украшения|Ювелирные изделия  
32                        Бытовая электроника|Ноутбуки  
227  Бытовая электроника|Аудио и видео|Телевизоры и...

Видим, что в бытовой технике могут быть довольно редкие сочетания букв, например "K50AB", есть слова с орфографическими ошибками ("гаршок"). Есть и частые слова, которые встретились в 2 их 6 объявлений.

**Названия категорий**

Посмотрим на названия категорий:

In [10]:
data.name.unique()[:3]

array(['Для дома и дачи|Посуда и товары для кухни|Посуда',
       'Хобби и отдых|Билеты и путешествия|Путешествия',
       'Личные вещи|Одежда, обувь, аксессуары|Женская одежда|Другое'],
      dtype=object)

Видим, что категории могут находиться на разном уровне иерархии. Посмотрим, насколько сильно может отличаться глубина категории:

In [11]:
data.name.apply(lambda x: len(x.split('|'))).value_counts().sort_index()

2     1406
3    15973
4     2621
Name: name, dtype: int64

Можно подумать, как использовать данные о том, что у классов есть иерархическая структура.

**Статистики по текстам**

Посмотрим, отличаются ли статистически тексты заголовков для разных категорий. Посчитаем количество уникальных символов в заголовке:

In [12]:
data.title.apply(len).describe()

count    20000.000000
mean        25.093300
std         12.474148
min          5.000000
25%         15.000000
50%         23.000000
75%         34.000000
max         50.000000
Name: title, dtype: float64

Посмотрим на на распределение количества слов:

In [13]:
data.title.apply(lambda x: len(x.split())).describe()

count    20000.00000
mean         3.79825
std          1.96368
min          1.00000
25%          2.00000
50%          3.00000
75%          5.00000
max         14.00000
Name: title, dtype: float64

В среднем в заголовках 3-4 слова. Заголовки короткие

Интересно посмотреть на краевые случаи. Давайте выведем примеры, где в заголовок получилось уместить больше 12 слов:

In [14]:
data[data.title.apply(lambda x: len(x.split()) > 12)].head(3)

title  category_id  \
1310  iPad pro 128 4G + клва + стилус + сумка и так. Д            3   
2613    HP 17.3 / Core i5 / 4 озу / 2 видео/ как новый            1   
3516   Отдам 2 альбома для фото 10 х 15 на 96 и 48 фот           19   

                                                   name  
1310  Бытовая электроника|Планшеты и электронные кни...  
2613                       Бытовая электроника|Ноутбуки  
3516  Для дома и дачи|Мебель и интерьер|Предметы инт...

Все примеры из бытовой электроники. Значит, количество слов может послужить предиктором категории. Посчитаем статистики для категорий:

In [15]:
data['word_count'] = data.title.apply(lambda x: len(x.split()))
data.groupby('name')['word_count'].mean().sort_values().head(3)

name
Личные вещи|Одежда, обувь, аксессуары|Женская одежда|Платья и юбки     2.576190
Для дома и дачи|Мебель и интерьер|Шкафы и комоды                       2.756219
Личные вещи|Одежда, обувь, аксессуары|Женская одежда|Верхняя одежда    2.824074
Name: word_count, dtype: float64

Действительно, среднее количество слов для категории отличается в 2 раза.

In [16]:
data.groupby('name')['word_count'].mean().sort_values().tail(3)

name
Хобби и отдых|Велосипеды|Запчасти и аксессуары            5.508600
Бытовая электроника|Телефоны|Аксессуары|Чехлы и плёнки    5.531429
Хобби и отдых|Билеты и путешествия|Путешествия            5.641399
Name: word_count, dtype: float64

Кажется, что статистики по тексту могут помочь улучшить качество работы классификатора. С другой стороны, нужно хорошенько подумать, стоит ли использовать такие признаки в сервисе?

In [17]:
data.shape

(20000, 4)

In [18]:
data_fixed = data.drop_duplicates(subset='title')

In [19]:
data_fixed.shape

(17925, 4)

## Бейзлайн

Разобьем данные на обучающую выборку и валидацию:

In [42]:
train_titles, val_titles, y_train, y_val = train_test_split(
    data['title'],
    data['category_id'],
    random_state=1, stratify=data['category_id']
)

Обычно, в задачах классификации первым бейзлайном является предсказание самым популярным классом. Но когда выборка сбалансирована, а классов много, такой бейзлайн не очень информативен:

In [25]:
(y_val == y_train.mode().values[0]).mean()

0.0224

## CountVectorizer

Для получения матриц признаков преобразуем заголовки с помощью CountVectorizer():

In [27]:
count_vectorizer = CountVectorizer()

X_train = count_vectorizer.fit_transform(train_titles.values)
X_val = count_vectorizer.transform(val_titles.values)

In [28]:
X_train

<15000x17309 sparse matrix of type '<class 'numpy.int64'>'
	with 54601 stored elements in Compressed Sparse Row format>

<p>В данном случае вместо&nbsp;<span class="math-tex">\(15000 \times 17316=259740000\)</span>&nbsp;значений мы храним всего&nbsp;<span class="math-tex">\(54676\)</span> тройки&nbsp;<span class="math-tex">\((i, j, x_j^{(i)})\)</span>. Это&nbsp;1500&nbsp;раз меньше!</p>

<p>Sparse формат не доставляет никаких неудобств &mdash;&nbsp;все полезные функции можно применять к матрицам в разреженном формате. Единственное &laquo;неудобство&raquo;&nbsp;&mdash;&nbsp;нельзя посмотреть&nbsp;на матрицу.<span style="color:#a03881"> </span>Но в данном случае этого делать не нужно, потому что всё, что мы увидим, это матрица&nbsp;из нулей. Не повторяйте этого на практике, на больших&nbsp;выборках преобразованные данные могут не поместиться в оперативную память:</p>

In [29]:
X_train.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

Посмотрим на получившийся словарь:

In [30]:
print('words in vocab: ', len(count_vectorizer.vocabulary_))
count_vectorizer.vocabulary_

words in vocab:  17309


{'samsung': 6097,
 'galaxy': 3857,
 's3': 6039,
 'duos': 3319,
 'фигурка': 16254,
 'китайского': 10554,
 'божка': 8106,
 'хотея': 16522,
 'из': 10093,
 'пласткомпозита': 13329,
 'санки': 14639,
 'коляска': 10738,
 'продается': 13878,
 'шкаф': 16910,
 'ортопедическая': 12817,
 'подушка': 13512,
 'для': 9507,
 'беременных': 7975,
 'кормящих': 10908,
 'продам': 13883,
 'шкуры': 16926,
 'енота': 9736,
 'бобра': 8091,
 'норки': 12567,
 'письменный': 13286,
 'стол': 15385,
 'компьютерный': 10817,
 'кроватки': 11080,
 'зверята': 9956,
 'владимир': 8625,
 'ретро': 14308,
 'свеча': 14722,
 'очень': 12953,
 'красивая': 11015,
 'телевизор': 15655,
 'omega': 5388,
 'tv': 6764,
 'светильник': 14699,
 'настольный': 12365,
 'роутер': 14435,
 'asus': 2303,
 'rt': 5992,
 'n18': 5205,
 'ковер': 10644,
 'шерстяной': 16857,
 'часы': 16675,
 'наручные': 12338,
 'minizhu': 5045,
 'sport': 6378,
 'коробке': 10914,
 'lenovo': 4650,
 'офисной': 12931,
 'работы': 14076,
 'ноутбук': 12583,
 'aser': 2285,
 'aspir

Убедимся, что необходимые значения хранятся по индексам из словаря:

In [31]:
count_vectorizer.transform(['Продам продам'])[0, count_vectorizer.vocabulary_['продам']]

2

Видим, что CountVectorizer автоматически приводит слова к нижнему регистру.

Важные параметры

Класс CountVectorizer содержит все нужные параметры для настройки преобразования под вашу задачу:

<pre>
<code class="language-python">CountVectorizer(
    *,
    input='content',
    encoding='utf-8',
    decode_error='strict',
    strip_accents=None,
    lowercase=True,
    preprocessor=None,
    tokenizer=None,
    stop_words=None,
    token_pattern='(?u)\\b\\w\\w+\\b',
    ngram_range=(1, 1),
    analyzer='word',
    max_df=1.0,
    min_df=1,
    max_features=None,
    vocabulary=None,
    binary=False,
    dtype=&lt;class 'numpy.int64'&gt;,
)</code></pre>

<p>Рассмотрим наиболее важные из них:</p>

<ul>
	<li><code>max_df</code>, <code>min_df</code>&nbsp;&mdash;&nbsp;параметры, позволяющие фильтровать слова по порогу встречаемости. Если вы передаёте в параметры действительные числа от 0 до 1, это это максимальная и минимальная доля встречаемости слова. Если целые, то абсолютное значение встречаемости;</li>
	<li><code>ngram_range</code>&nbsp;&mdash;&nbsp;диапазон n-грамм: (1, 3) &mdash;&nbsp;значит, включить в словарь&nbsp;униграммы, биграммы и тригаммы;</li>
	<li><code>analyzer</code>&nbsp;&mdash;&nbsp;определяет, что считать токеном. Есть возможность токенезировать по словам, буквам и буквам с учётом окончаний слов. Может работать особенно эффективно в сочетании с учётом n-грамм;</li>
	<li><code>token_pattern</code>&nbsp;&mdash; регулярное выражение для токенизации.</li>
</ul>


## Модель логистической регрессии

Обучим модель логистической регрессии c параметрами по умолчанию и посчитаем accuracy:

In [32]:
model = LogisticRegression()
model.fit(X_train, y_train)

preds = model.predict(X_val)
accuracy_score(preds, y_val)

0.7412

In [33]:
X_train.shape

(15000, 17309)

Логистическая регрессия позволяет получить не только предсказанные категории, но и вероятности. Предсказания для одного объекта в нашем случае выглядят как вектор из 54 чисел:

In [34]:
np.round(model.predict_proba(X_val[0]), 3)

array([[0.   , 0.003, 0.02 , 0.004, 0.   , 0.951, 0.   , 0.   , 0.   ,
        0.001, 0.   , 0.   , 0.001, 0.002, 0.   , 0.   , 0.   , 0.   ,
        0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.006,
        0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.001, 0.   , 0.   ,
        0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   , 0.   ,
        0.   , 0.   , 0.   , 0.   , 0.001, 0.   , 0.   , 0.   , 0.   ]])

Проверим свою интуицию по поводу весов полученной модели. Мы знаем, что слова встречающиеся в разных категориях, не должны получить большой вес в модели. Посмотрим на вес слова 'продам' для разных категорий:

In [35]:
model.coef_[:, count_vectorizer.vocabulary_['продам']]

array([ 0.34559004,  0.06288177,  0.59957824,  0.13807027,  0.10687667,
       -0.39806094,  0.23464759,  0.63089573,  0.23181878, -0.89111716,
        0.83826997,  0.29862008,  0.02971731, -0.4415758 ,  0.036928  ,
       -0.17990815,  0.69849329, -0.73701999,  0.51683897, -0.53060521,
       -0.71550723,  0.32470807,  0.38634688,  0.09264987, -0.50870911,
        0.30705982,  0.6656596 ,  0.12929302,  0.20823902,  0.55326054,
       -0.4158767 ,  0.29060007,  0.46290202,  1.05240447,  0.26684442,
        0.24532754, -0.48517243, -0.04419311,  0.48383919,  0.18881584,
       -0.48806713,  0.21480194, -0.26896261, -1.02754316, -0.43948847,
       -0.24825822,  0.31478717, -0.37006392, -1.33041696,  0.22437298,
        0.56794611, -0.95311163, -0.34761108, -0.92781626])

Слова, по которым можно определить категорию, наоборот, должны получить большой вес. Посмотрим на вес для слова 'iphone' для разных категорий:

In [36]:
model.coef_[:, count_vectorizer.vocabulary_['iphone']]

array([ 6.25565865,  0.26212449, -0.17869003, -0.39511054, -0.18325014,
       -0.17473601,  1.38369292, -0.16828429, -0.23135096,  3.27542748,
       -0.17442942, -0.12760618,  0.16009538, -0.15325199, -0.31435685,
       -0.41114944, -0.14462141, -0.3379351 , -0.20750758, -0.30443582,
       -0.33418348, -0.3254551 , -0.17350491, -0.38266472, -0.34648044,
       -0.12081226, -0.18453041, -0.304368  , -0.21939205, -0.27569734,
       -0.33228105,  0.34898225, -0.19440275, -0.17996807, -0.20384138,
       -0.12443814, -0.16540909, -0.39304052, -0.31431479, -0.21566646,
       -0.30704842, -0.24220425, -0.1792247 , -0.32836335, -0.35204034,
       -0.21427822,  0.24439209, -0.39829508, -0.19034804, -0.35220215,
       -0.21021343, -0.25817748, -0.29134197, -0.30946912])

Действительно, вес при этом слове для некоторых категорий очень большой. Ведь если это слово есть в заголовке, скорее всего, правильной категорией будет что-то начинающееся на 'Бытовая электроника|Телефоны|'. Проверим, так ли это:

In [37]:
cat_names = data.set_index('category_id')['name'].to_dict()
top_iphone_cats = np.argsort(model.coef_[:, count_vectorizer.vocabulary_['iphone']])[-3:]
[cat_names[c] for c in top_iphone_cats]

['Бытовая электроника|Телефоны|Другие марки',
 'Бытовая электроника|Телефоны|Аксессуары|Чехлы и плёнки',
 'Бытовая электроника|Телефоны|iPhone']

**Ошибки модели**

Посмотрим, в каких категориях чаще всего происходят ошибки. Для этого создадим удобный dataframe:

In [38]:
models_erros = pd.DataFrame({
    'title': val_titles.values,
    'preds_category': [cat_names[cat] for cat in preds],
    'true_category': [cat_names[cat] for cat in y_val],
    'preds_ok': preds == y_val
})

Выведем 3 случайные ошибки модели:

In [39]:
models_erros[~models_erros['preds_ok']].sample(10)

title  \
11698                    Электрический набор Комплекты   
17099                            Продам часы и рубашку   
14596                 Шпилька оцинкованная М8 длина 2м   
8116                                Итальянские сапоги   
12935             Xbox one с двумя геймпадами и играми   
9456                                     Столик и стул   
14110     Флористическая композиция " Корзина счастья"   
6108                       Дисплей от Samsung R425 14"   
1334                                      Рико Матрикс   
17665  Эльбрус (сувенир капилка) бизнес, лыжи сноуборд   

                                          preds_category  \
11698  Для дома и дачи|Бытовая техника|Для кухни|Мелк...   
17099                  Личные вещи|Часы и украшения|Часы   
14596  Для дома и дачи|Ремонт и строительство|Инструм...   
8116   Личные вещи|Одежда, обувь, аксессуары|Женская ...   
12935  Бытовая электроника|Игры, приставки и программ...   
9456    Для дома и дачи|Мебель и интерьер|Столы и стулья   
14110                 Хобби и отдых|Спорт и отдых|Другое   
6108                Бытовая электроника|Телефоны|Samsung   
1334   Для дома и дачи|Ремонт и строительство|Стройма...   
17665     Хобби и отдых|Спорт и отдых|Зимние виды спорта   

                                           true_category  preds_ok  
11698     Хобби и отдых|Велосипеды|Запчасти и аксессуары     False  
17099            Хобби и отдых|Коллекционирование|Другое     False  
14596  Для дома и дачи|Ремонт и строительство|Стройма...     False  
8116    Личные вещи|Одежда, обувь, аксессуары|Аксессуары     False  
12935  Бытовая электроника|Игры, приставки и программ...     False  
9456   Личные вещи|Товары для детей и игрушки|Детская...     False  
14110  Для дома и дачи|Мебель и интерьер|Предметы инт...     False  
6108   Бытовая электроника|Товары для компьютера|Мони...     False  
1334   Личные вещи|Товары для детей и игрушки|Детские...     False  
17665                 Хобби и отдых|Спорт и отдых|Другое     False

Видим, что в первых двух строчках ошибки, скорее всего, произошли по причине опечаток. Нужно подумать над тем, как повысить устойчивость модели к опечаткам и ошибкам. Этим вам предстоит заняться в домашнем задании.

На этом анализ ошибок, конечно же, заканчивать не стоит — но мы не будем продолжать это делать здесь, а оставим в виде упражнения.

## Word2vec

Импортируем gensim – библиотеку, с помощью которой будем обучать модель w2v. Также создадим понадобятся вспомогательные классы, которые будут логировать ошибку после каждой итерации обучения.

In [29]:
import re
import gensim
from gensim.models.word2vec import Word2Vec
from gensim.models.fasttext import FastText
from gensim.models.callbacks import CallbackAny2Vec

class LossLogger(CallbackAny2Vec):
    def __init__(self):
        self.epoch = 0
        self.loss_previous_step = 0

    def on_epoch_end(self, model):
        loss = model.get_latest_training_loss()
        if self.epoch == 0:
            print('Loss after epoch {}: {}'.format(self.epoch, loss))
        else:
            print('Loss after epoch {}: {}'.format(self.epoch, loss - self.loss_previous_step))
        self.epoch += 1
        self.loss_previous_step = loss

class EpochLogger(CallbackAny2Vec):
    def __init__(self):
        self.epoch = 0

def on_epoch_end(self, model):
    print(f'Epoch {self.epoch}')
    self.epoch += 1

print(gensim.__version__)

4.3.2


Прочитаем данные и выведем первые несколько строк:

In [27]:
w2v_train_data = pd.read_csv(way + 'unlabeled_data.csv',  index_col=0)
print(w2v_train_data.shape)
w2v_train_data.head()

(200000, 2)


item_id                                              title
30000   377939                            1 рубль царские 10 штук
30001   384779  Напольный насос giyo высокого давления, с мано...
30002   276889                                Sony Xperia m2 Dual
30003   155718   Zoom B2.1U+ AC Процессор эффектов для бас-гитары
30004   260705                    Люстра Favourite petite 1574-7P

Данные для обучения word2vec — это просто тексты или любые другие наборы последовательностей токенов, для которых можно предположить справедливость дистрибутивной гипотезы. Токенизируем тексты:

In [30]:
WORD_PATTERN = '(?u)\\b\\w\\w+\\b'  # паттерн для токенизации возьмем из векторайзеров sklearn
reg_exp = re.compile(pattern=WORD_PATTERN)  # скомпилируем регулярное выражение
sentences = [reg_exp.findall(s.lower()) for s in w2v_train_data.title]

Выведем результат токенизации:

In [31]:
sentences[:3]

[['рубль', 'царские', '10', 'штук'],
 ['напольный', 'насос', 'giyo', 'высокого', 'давления', 'манометр'],
 ['sony', 'xperia', 'm2', 'dual']]

Обучим w2v с параметрами по усмолчанию:

In [44]:
w2v_model = Word2Vec()
w2v_model.build_vocab(sentences)
w2v_model.train(
   corpus_iterable=sentences,
   total_examples=w2v_model.corpus_count,
   epochs=30,
   compute_loss=True,
   callbacks=[LossLogger()]
)

Loss after epoch 0: 766606.25
Loss after epoch 1: 440904.875
Loss after epoch 2: 358823.75
Loss after epoch 3: 309357.875
Loss after epoch 4: 307826.0
Loss after epoch 5: 275987.0
Loss after epoch 6: 250227.75
Loss after epoch 7: 248417.75
Loss after epoch 8: 227534.25
Loss after epoch 9: 219693.5
Loss after epoch 10: 228336.0
Loss after epoch 11: 198833.75
Loss after epoch 12: 192022.5
Loss after epoch 13: 190561.25
Loss after epoch 14: 152454.0
Loss after epoch 15: 149213.5
Loss after epoch 16: 137058.5
Loss after epoch 17: 143558.5
Loss after epoch 18: 147469.0
Loss after epoch 19: 144050.0
Loss after epoch 20: 141599.5
Loss after epoch 21: 141346.0
Loss after epoch 22: 139416.0
Loss after epoch 23: 143226.5
Loss after epoch 24: 141358.0
Loss after epoch 25: 145964.5
Loss after epoch 26: 135681.0
Loss after epoch 27: 133504.5
Loss after epoch 28: 127210.0
Loss after epoch 29: 129034.5


(17445628, 22027110)

Попытаемся понять, получилось ли обучиться. Выведем векторы, близкие к вектору слова 'продам';

In [45]:
w2v_model.wv.similar_by_word('продам')

[('продаю', 0.9066060185432434),
 ('породам', 0.7147958874702454),
 ('отдам', 0.577515184879303),
 ('продаются', 0.550401508808136),
 ('продается', 0.5097194910049438),
 ('новую', 0.4678470194339752),
 ('продаём', 0.46553313732147217),
 ('продаётся', 0.45866894721984863),
 ('подам', 0.4399629533290863),
 ('продадим', 0.4269896149635315)]

Посмотрим на ещё одно слово. Интересно, что можно поискать слова близкие к слову 'айфон' и тем самым получить написание этого слова на английском:

In [46]:
w2v_model.wv.similar_by_word('айфон')

[('iphone', 0.8191383481025696),
 ('iphone5', 0.7969886064529419),
 ('iphone4', 0.7495058178901672),
 ('5c', 0.7292015552520752),
 ('iphon', 0.7287685871124268),
 ('5с', 0.7111333012580872),
 ('айфон4', 0.7040383219718933),
 ('iphone6', 0.6889973878860474),
 ('ipone', 0.67149418592453),
 ('4s', 0.6342318058013916)]

Напишем класс, которые позволит там получать векторы для заголовков, как сумму векторов входящих в него слов:

In [ ]:
class Word2VecTransformer:

    def __init__(self, w2v_model, word_pattern):

        self.w2v_model = w2v_model
        self.word_pattern = word_pattern
        self.re = re.compile(pattern=self.word_pattern)

    def fit(self, X):
        return self

    def transform(self, X):

        X_transformed = np.zeros((len(X), self.w2v_model.wv.vector_size))
        for i, title in enumerate(X):

            title_vector = np.zeros((self.w2v_model.wv.vector_size,))
            tokens = self.re.findall(title.lower())
            if len(tokens) == 0:
              continue

            for token in tokens:
                if token in self.w2v_model.wv.key_to_index:
                    title_vector += self.w2v_model.wv.get_vector(token)

            X_transformed[i] = title_vector / len(tokens)

        return X_transformed

Трансформируем с помощью него обучающую выборку и обучим логистическую регрессию:

In [48]:
w2v_transformer = Word2VecTransformer(w2v_model=w2v_model, word_pattern=WORD_PATTERN)

train_w2v = w2v_transformer.transform(train_titles.values)
val_w2v = w2v_transformer.transform(val_titles.values)

model_on_w2v = LogisticRegression(solver='liblinear')
model_on_w2v.fit(train_w2v, y_train)

LogisticRegression(solver='liblinear')

Оценим качество классификатора:

In [49]:
preds = model_on_w2v.predict(val_w2v)
accuracy_score(preds, y_val)

0.727

Варианты улучшений:

* Word2vec не может получить вектора слов, которых не было при обучении модели. Однако есть модель, очень похожая на word2vec, но использующая при обучении буквенные n-гаммы. Эта модель называется Fasttext — в библиотеке gensim есть её реализация;
* Метод агрегации, который мы использовали, очень грубый: можно попробовать не суммировать, а считать среднее или суммировать с весами;
* Можно попробовать подобрать более оптимальные гиперпараметры w2v: использовать skipgram вместо CBOW, учить с другим размером окна, использовать другой размер вектора, изменить соотношение позитивных и негативных примеров в negative sampling или использовать иерархический софтмакс.

# Поехали

In [20]:
class Word2VecTransformer:

    def __init__(self, w2v_model, word_pattern):

        self.w2v_model = w2v_model
        self.word_pattern = word_pattern
        self.re = re.compile(pattern=self.word_pattern)

    def fit(self, X):
        return self

    def transform(self, X):

        X_transformed = np.zeros((len(X), self.w2v_model.wv.vector_size))
        for i, title in enumerate(X):

            title_vector = np.zeros((self.w2v_model.wv.vector_size,))
            tokens = self.re.findall(title.lower())
            if len(tokens) == 0:
              continue

            for token in tokens:
                if token in self.w2v_model.wv.key_to_index:
                    title_vector += self.w2v_model.wv.get_vector(token)

            X_transformed[i] = title_vector / len(tokens)

        return X_transformed

In [21]:
def accuracy3(probs):
  top3_classes = np.argsort(probs, axis=1)[:, -3:]
  correct_predictions = [y_val.iloc[i] in top3_classes[i] for i in range(len(y_val))]
  return np.mean(correct_predictions)

In [22]:
def ft_eval(ft_model_gr):
  ft_transformer = Word2VecTransformer(w2v_model=ft_model_gr,
                                       word_pattern=WORD_PATTERN)

  train_ft = ft_transformer.transform(train_titles.values)
  val_ft = ft_transformer.transform(val_titles.values)

  model_on_ft = LogisticRegression(solver='liblinear')
  model_on_ft.fit(train_ft, y_train)

  return accuracy3(model_on_ft.predict_proba(val_ft))

In [ ]:
from gensim.models import FastText

ft_model = FastText(vector_size=250, window=5, min_count=4)
ft_model.build_vocab(sentences)
ft_model.train(
    corpus_iterable=sentences,
    total_examples=ft_model.corpus_count,
    epochs=30,
    compute_loss=True,
    callbacks=[LossLogger()]
)

In [60]:
ft_model.wv.similar_by_word('продам')

[('продаю', 0.9847356677055359),
 ('продаём', 0.9560081958770752),
 ('продадим', 0.9543347358703613),
 ('продаем', 0.9258226156234741),
 ('продажу', 0.9176304340362549),
 ('продать', 0.9151418805122375),
 ('продаже', 0.9087103009223938),
 ('продаются', 0.9052717089653015),
 ('продаж', 0.9047023057937622),
 ('продаеться', 0.9019080996513367)]

### для ворд2века

In [70]:
class_probs = model_on_w2v.predict_proba(val_w2v)
top3_classes = np.argsort(class_probs, axis=1)[:, -3:]
top3_classes

array([[24, 15, 31],
       [ 6,  1,  3],
       [ 9,  6, 10],
       ...,
       [47, 51, 43],
       [23, 21, 15],
       [44, 47, 53]])

In [79]:
correct_predictions = [y_val.iloc[i] in top3_classes[i] for i in range(len(y_val))]
np.mean(correct_predictions)

0.8576

### для фасттекста

In [ ]:
#{'vector_size': 100, 'window': 5, 'min_count': 1, 'epochs': 30, 'sg': 1}
#{'vector_size': 200, 'window': 7, 'min_count': 1, 'epochs': 20, 'sg': 1}
{'vector_size': 200, 'window': 7, 'min_count': 1, 'epochs': 20, 'sg': 1}

In [59]:
from gensim.models import FastText

ft_model = FastText(vector_size=300, window=6, min_count=1, sg = 1)
ft_model.build_vocab(sentences)
ft_model.train(
    corpus_iterable=sentences,
    total_examples=ft_model.corpus_count,
    epochs=30,
    compute_loss=True,
    #callbacks=[LossLogger()]
)

Loss after epoch 0: 0.0
Loss after epoch 1: 0.0
Loss after epoch 2: 0.0
Loss after epoch 3: 0.0
Loss after epoch 4: 0.0
Loss after epoch 5: 0.0
Loss after epoch 6: 0.0
Loss after epoch 7: 0.0
Loss after epoch 8: 0.0
Loss after epoch 9: 0.0
Loss after epoch 10: 0.0
Loss after epoch 11: 0.0
Loss after epoch 12: 0.0
Loss after epoch 13: 0.0
Loss after epoch 14: 0.0
Loss after epoch 15: 0.0
Loss after epoch 16: 0.0
Loss after epoch 17: 0.0
Loss after epoch 18: 0.0
Loss after epoch 19: 0.0


(11914799, 14684740)

In [50]:
ft_transformer = Word2VecTransformer(w2v_model=ft_model, word_pattern=WORD_PATTERN)

train_ft = ft_transformer.transform(train_titles.values)
val_ft = ft_transformer.transform(val_titles.values)

model_on_ft = LogisticRegression(solver='liblinear')
model_on_ft.fit(train_ft, y_train)

LogisticRegression(solver='liblinear')

In [51]:
accuracy3(model_on_ft.predict_proba(val_ft))

0.9072

### Грид серч

In [ ]:
param_grid = {
    'vector_size': [100, 200, 300],
    'window': [3, 5, 7],
    'min_count': [1, 3, 5, 7],
    'epochs': [20, 30]
}

In [ ]:
#{'vector_size': 100, 'window': 5, 'min_count': 1, 'epochs': 30, 'sg': 1}
#{'vector_size': 200, 'window': 7, 'min_count': 1, 'epochs': 20, 'sg': 1}
#{'vector_size': 200, 'window': 7, 'min_count': 1, 'epochs': 20, 'sg': 1}
{'vector_size': 200, 'window': 9, 'min_count': 2, 'epochs': 30, 'sg': 1}

In [70]:
from gensim.models import FastText
import itertools

# Определение параметров для перебора
param_grid = {
    'vector_size': [100, 200],
    'window': [3, 5, 7],
    'min_count': [1, 3, 5, 7],
    'epochs': [20, 30],
    'sg': [0, 1]
}

# Итерация по всем комбинациям параметров
best_score = None
best_params = None

for params in itertools.product(*param_grid.values()):
    vector_size, window, min_count, epochs, sg = params

    # Создание и обучение модели
    ft_model = FastText(vector_size=vector_size, window=window, min_count=min_count, sg=sg)
    ft_model.build_vocab(sentences)
    ft_model.train(
        corpus_iterable=sentences,
        total_examples=ft_model.corpus_count,
        epochs=epochs,
        compute_loss=True
        # Вы можете добавить здесь callbacks, если они вам нужны
    )

    # Оценка модели
    score = ft_eval(ft_model)

    # Сохранение лучших параметров и результата
    if best_score is None or score > best_score:
        best_score = score
        best_params = {
            'vector_size': vector_size,
            'window': window,
            'min_count': min_count,
            'epochs': epochs,
            'sg': sg
        }
        print(best_score)
        print(best_params)

print("Лучший результат:", best_score)
print("Лучшие параметры:", best_params)


0.8524
{'vector_size': 100, 'window': 3, 'min_count': 1, 'epochs': 20, 'sg': 0}
0.8904
{'vector_size': 100, 'window': 3, 'min_count': 1, 'epochs': 20, 'sg': 1}
0.8952
{'vector_size': 100, 'window': 3, 'min_count': 1, 'epochs': 30, 'sg': 1}
0.8982
{'vector_size': 100, 'window': 5, 'min_count': 1, 'epochs': 30, 'sg': 1}


KeyboardInterrupt: ignored

## Оптуна

In [32]:
import optuna
from gensim.models import FastText

def objective(trial):
    # Определение диапазона параметров
    vector_size = trial.suggest_categorical('vector_size', [200, 250, 300])
    window = trial.suggest_int('window', 6, 10)
    #min_count = trial.suggest_int('min_count', 1, 6)
    #epochs = trial.suggest_categorical('epochs', [20, 30])
    #sg = trial.suggest_categorical('sg', [0, 1])

    # Создание и обучение модели
    ft_model = FastText(vector_size=vector_size, window=window, min_count=1, sg = 1)
    ft_model.build_vocab(sentences)
    ft_model.train(
        corpus_iterable=sentences,
        total_examples=ft_model.corpus_count,
        epochs=30
    )

    # Оценка модели
    score = ft_eval(ft_model)
    return score

study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=15)  # Укажите количество испытаний

print("Лучший результат:", study.best_value)
print("Лучшие параметры:", study.best_params)

[I 2023-11-20 18:08:27,656] A new study created in memory with name: no-name-3e516208-84e4-45be-8ab4-7168ab81e69f
[I 2023-11-20 18:14:50,196] Trial 0 finished with value: 0.9092 and parameters: {'vector_size': 200, 'window': 10}. Best is trial 0 with value: 0.9092.
[I 2023-11-20 18:22:15,238] Trial 1 finished with value: 0.91 and parameters: {'vector_size': 300, 'window': 6}. Best is trial 1 with value: 0.91.
[I 2023-11-20 18:29:56,198] Trial 2 finished with value: 0.908 and parameters: {'vector_size': 300, 'window': 9}. Best is trial 1 with value: 0.91.
[I 2023-11-20 18:37:20,467] Trial 3 finished with value: 0.9048 and parameters: {'vector_size': 250, 'window': 6}. Best is trial 1 with value: 0.91.
[I 2023-11-20 18:43:31,539] Trial 4 finished with value: 0.909 and parameters: {'vector_size': 200, 'window': 10}. Best is trial 1 with value: 0.91.
[I 2023-11-20 18:51:00,205] Trial 5 finished with value: 0.906 and parameters: {'vector_size': 250, 'window': 6}. Best is trial 1 with value:

KeyboardInterrupt: ignored

In [53]:
test_transformer = Word2VecTransformer(w2v_model=fin_model,
                                       word_pattern=WORD_PATTERN)

train_ft = test_transformer.transform(train_titles.values)
val_ft = test_transformer.transform(val_titles.values)

In [ ]:
import optuna
from gensim.models import FastText


def objective(trial):
    # Определение диапазона параметров
    c = trial.suggest_float('c', 1e-10, 1e10, log=True)
    #solver = trial.suggest_categorical('solver', ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'])

    # Создание и обучение модели
    model_test = LogisticRegression(solver="liblinear", C=c)
    model_test.fit(train_ft, y_train)

    # Оценка модели
    score = accuracy3(model_test.predict_proba(val_ft))
    return score

study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=15)  # Укажите количество испытаний

print("Лучший результат:", study.best_value)
print("Лучшие параметры:", study.best_params)

[I 2023-11-20 20:19:54,074] A new study created in memory with name: no-name-25e13088-443f-4132-bae5-fcb7e41d0c4e
[I 2023-11-20 20:20:35,897] Trial 0 finished with value: 0.8798 and parameters: {'c': 0.004092009664823455}. Best is trial 0 with value: 0.8798.
[I 2023-11-20 20:21:01,174] Trial 1 finished with value: 0.7674 and parameters: {'c': 0.00011794270755526602}. Best is trial 0 with value: 0.8798.


# Финал

In [24]:
df_test = pd.read_csv(way + "suggest_test.csv")

In [25]:
X_train_fin = data['title']
y_train_fin = data['category_id']

X_test_fin = df_test['title']

провери:
vector_size=300, window=6, min_count=1, sg = 1

In [32]:
from gensim.models import FastText

fin_model = FastText(vector_size=300, window=10, min_count=1, sg = 1)
fin_model.build_vocab(sentences)
fin_model.train(
    corpus_iterable=sentences,
    total_examples=fin_model.corpus_count,
    epochs=30,
    compute_loss=True,
    #callbacks=[LossLogger()]
)

(20888134, 22027110)

In [33]:
fin_transformer = Word2VecTransformer(w2v_model=fin_model, word_pattern=WORD_PATTERN)

train_fin = fin_transformer.transform(X_train_fin.values)
test_fin = fin_transformer.transform(X_test_fin.values)

model_on_fin = LogisticRegression(solver='liblinear')
model_on_fin.fit(train_fin, y_train_fin)

LogisticRegression(solver='liblinear')

In [34]:
class_probs = model_on_fin.predict_proba(test_fin)
top3_classes = np.argsort(class_probs, axis=1)[:, -3:]
top3_classes

array([[26,  2,  5],
       [38, 31, 30],
       [27, 23, 21],
       ...,
       [34, 30, 32],
       [47, 17, 26],
       [44, 19, 15]])

In [35]:
pd.DataFrame(top3_classes).to_csv('solution.csv', header=['top1', 'top2', 'top3'], index=False)